### ISSUE TITLE
Final SPL in `filter_noise` in `defense/boost.py`

### ISSUE THREAD

**Issue by skoffas** [charlesjin/compatibility_clustering#3]
Hi again,

I am sorry for the spam but I have another question.

In the end of `filter_noise` function in `defense/boost.py` there is an SPL loop that trains the model with the clean samples and every 5 epochs the clean set is updated. In Appendix C in the paper the following is mentioned: 
a) "For CIFAR-10, we also include a final self-training step by training a fresh model for 100 epochs on the recovered samples"
b) "Therefore, we use a simple self-paced learning algorithm to dynamically adjust the samples during training to limit the self-training to a single iteration". 

I have run some initial experiments with MNIST and it seems that when in the clean set before this SPL loop there are some poisoned samples there is a chance that during this refining more poisoned samples are selected as the trained model learns the backdoor and the loss decreases a lot. In one of my experiments I used 5% poisoned data (i.e., 3000 samples for MNIST) and even though before this step the algorithm has discarded most of them (around 2600) after the SPL all of them were selected again. 

My question is the following: In order to use your defense correctly, do you think it is better to disable this step for all my experiments, enable it only for CIFAR10 (if I use it), or something else?

Thanks for you time and help.

**Comment by charlesjin** [comment]
Can you clarify what model you are using and how many epochs?

The final SPL step may not be necessary if the accuracy is sufficient without the additional training.

**Comment by skoffas** [comment]
I have embedded the defense code in my pipeline without any changes so the final SPL step is included in the `filter_noise` function. I did not change anything in the SPL loop so it uses 100 epochs (20 * 5). The model architecture I am using is Resnet18 with the MNIST dataset. After SPL finishes I am using the `keep` list to train a model from scratch (because it is MNIST just 30 epochs are enough for a good performance) with the identified clean samples. For MNIST the accuracy is quite high so I think that SPL is not required. In general, it may be better to run all my experiments (even with other datasets) without the final SPL step and and observe the accuracy that the models have after cleaning the data. Then I can test again with SPL only the cases that the accuracy decreases a lot (more than 10%). What do you think?

In our implementation there is a difference in the way we poison the data. In this repo the poisoning always uses a source class for the poisoned samples (source-specific backdoor attack), but I am not using a specific source class (source-agnostic backdoor attack). In the paper in the threat model the only requirement mentioned for this attack is that we cannot poison more than 50% of the data, so I thought that I can also use a source-agnostic backdoor. Do you think that this may result in a different behavior by the defense?

**Comment by charlesjin** [comment]
For MNIST, ResNet18 with 100 epochs is probably overkill. You may need a network with less capacity + reduce the number of epochs. Right now, the network is probably able to fit both backdoors AND real signals in the initial weak learning phase. Your idea to leave out the SPL step at the end makes sense.

We actually evaluated on source-agnostic attacks and it works the same.

**Comment by skoffas** [comment]
Everything is clear now. I will leave it our for now as I observed that the weak learner learns both backdoors and real signals hurting the defense's performance.

Thanks for the reply and your time. You are free to close this issue.
